### DISCUSSION TITLE
Should this code which passes dict as separate args type check correctly?

### DISCUSSION THREAD

**Discussion by AlexeyDmitriev** [microsoft/pyright#7460] [General]
So, I have the following code:

```python
def f(x: int, y: str):
    pass


def g():
    args = {'x': 3, 'y': "5"} # we use args here instead of passing directly so that we can use it twice
    print(args)
    f(**args)
```

it doesn't typecheck correctly, because args is detected as `dict[str, int | str]` which isn't necessarily OK to pass to `f`.
However, it's a valid code.
We can also show to pyright that it's valid by adding annotation like this:

```python
def g():
    class CurrentDict(typing.TypedDict):
        x: int
        y: str

    args: CurrentDict = {'x': 3, 'y': "5"}
    print(args)
    f(**args)
```

Which feels unnecessary/too wordy.

Is it possible to make pyright understand that this code is fine without explicit annotation?
I understand that in general you can't treat `TypedDict` as `dict`, but maybe we can understand here that it is `both` (i.e whichever works is fine)

**Comment by erictraut** [comment]
What you're suggesting would not be type safe and would violate the type specification, so that's not something we'd implement in pyright.

There's really no working around the need to define a TypedDict here. 

There have been some proposals to introduce a lighter-weight way to define a TypedDict inline. Pyright has experimental support for one such proposal, but this hasn't gone beyond the discussion phase. If you're interested in formally proposing something like this, you could write a PEP and shepherd it through the standardization process.

```python
# pyright: enableExperimentalFeatures=true

def f(x: int, y: str):
    pass

def g():
    args: dict[{"x": int, "y": str}] = {"x": 3, "y": "5"}
    print(args)
    f(**args)
```

**Reply by AlexeyDmitriev** [reply]
I think the experimental solution you mentioned doesn't help much, you still have to write types explicitly and repeat the names.

I guess something like would work but it's too adhoc:
```
args: TreatItAsTypeDictNotPlainDict = {"x": 3, "y": "5"}
```

> What you're suggesting would not be type safe and would violate the type specification

I believe that it will violate the current specification, but can you clarify what do you mean by "would not be typesafe"?
As I mentioned, I understand that treating TypedDict as dict in general is not possible, but I thought maybe it's possible to do something along the lines "try interpreting this if it was TypedDict if it works, OK, try interpreting this as dict if it works also"  which is type safe, but of course have computation complexity problem

**Reply by erictraut** [reply]
The rules for subtyping can't be as loose as "try it multiple ways and see which one works". A type checker needs to follow strict rules about type compatibility. This is the key to type safety and the composability of typing features.

There are good reasons why a TypedDict cannot (generally speaking) be compatible with a dict. These are discussed in detail in [PEP 589](https://peps.python.org/pep-0589/), which introduced TypedDict.

You may find some of the information in [draft PEP 728 ](https://peps.python.org/pep-0728/) interesting to read.
